core/src/main/java/org/jdbi/v3/core/mapper/reflect/internal/ImmutablesPropertiesFactory.java

```diff
@@ -0,0 +1,266 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.core.mapper.reflect.internal;
+
+import java.lang.annotation.Annotation;
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import net.jodah.expiringmap.ExpirationPolicy;
+import net.jodah.expiringmap.ExpiringMap;
+
+import org.jdbi.v3.core.generic.GenericTypes;
+import org.jdbi.v3.core.internal.exceptions.Unchecked;
+import org.jdbi.v3.core.mapper.reflect.internal.PojoProperties.PojoProperty;
+import org.jdbi.v3.core.qualifier.QualifiedType;
+import org.jdbi.v3.core.qualifier.Qualifiers;
+
+public class ImmutablesPropertiesFactory {
+
+    private static final Map<Type, PojoProperties<?>> IMMUTABLE_PROPERTIES = ExpiringMap
+            .builder()
+            .expiration(10, TimeUnit.MINUTES)
+            .expirationPolicy(ExpirationPolicy.ACCESSED)
+            .build();
+
+    private static final Map<Type, PojoProperties<?>> MODIFIABLE_PROPERTIES = ExpiringMap
+            .builder()
+            .expiration(10, TimeUnit.MINUTES)
+            .expirationPolicy(ExpirationPolicy.ACCESSED)
+            .build();
+
+    private ImmutablesPropertiesFactory() {}
+
+    @SuppressWarnings("unchecked")
+    public static <T, B> Function<Type, PojoProperties<T>> immutable(Class<T> defn, Supplier<B> builder) {
+        return t -> (PojoProperties<T>) IMMUTABLE_PROPERTIES.computeIfAbsent(t, x -> new ImmutablePojoProperties<>(t, defn, builder));
+    }
+
+    @SuppressWarnings("unchecked")
+    public static <T, M extends T> Function<Type, PojoProperties<T>> modifiable(Class<T> defn, Supplier<M> constructor) {
+        return t -> (PojoProperties<T>) MODIFIABLE_PROPERTIES.computeIfAbsent(t, x -> new ModifiablePojoProperties<>(t, defn, constructor));
+    }
+
+    static MethodHandle alwaysSet() {
+        return MethodHandles.dropArguments(MethodHandles.constant(boolean.class, true), 0, Object.class);
+    }
+
+    abstract static class BasePojoProperties<T, B> extends PojoProperties<T> {
+        protected final Map<String, ImmutablesPojoProperty<T>> properties = new HashMap<>();
+        protected final Class<?> defn;
+        protected final Supplier<?> builder;
+
+        BasePojoProperties(Type type, Class<?> defn, Supplier<B> builder) {
+            super(type);
+            this.defn = defn;
+            this.builder = builder;
+            for (Method m : defn.getMethods()) {
+                if (isProperty(m)) {
+                    final String name = m.getName();
```

Comment:
Immutables supports both `getName()` and `name()` method naming patterns out of the box--we should probably run this value through a function that strips the `get`/`is` prefix if present.